File: backend/app/api/plans.py
from typing import Literal

from fastapi import APIRouter, Depends, HTTPException, status
from motor.motor_asyncio import AsyncIOMotorDatabase

from app.core.db import get_db
from app.core.security import get_current_user
from app.models.job import JobResponse
from app.models.plan import (
    InjuryReport,
    PlanOverview,
    PlanProgress,
    PlanRequest,
    PlanResponse,
    PlanVersionSummary,
    SessionDeleteRequest,
    SessionDurationRequest,
    SessionLinkInfo,
    SessionLinkRequest,
    SessionMoveRequest,
    SessionSkipRequest,
    TodaySession,
    Weekday,
    WeekReconciliation,
)
from app.services import (
    plan_completion_service,
    plan_moves_service,
    plan_progress,
    plan_reconcile_service,
    plan_service,
)

router = APIRouter(prefix="/api/v1/plans", tags=["plans"])


def _no_plan_error() -> HTTPException:
    return HTTPException(
        status_code=status.HTTP_409_CONFLICT,
        detail={"code": "NO_PLAN", "message": "Aucun plan actif."},
    )


def _session_not_found_error() -> HTTPException:
    return HTTPException(
        status_code=status.HTTP_404_NOT_FOUND,
        detail={"code": "SESSION_NOT_FOUND", "message": "Séance introuvable pour ce jour."},
    )


@router.post("", response_model=JobResponse, status_code=status.HTTP_202_ACCEPTED)
async def create_plan(
    body: PlanRequest,
    db: AsyncIOMotorDatabase = Depends(get_db),
    user: dict = Depends(get_current_user),
):
    """Queue a plan generation and return its job.

    Not synchronous, and it can't be: a 17-week plan is ~12.5k output tokens,
    three to six minutes of model time. Poll `GET /api/v1/jobs/{id}`; once it
    reports `done`, `GET /api/v1/plans/current` has the plan."""
    return await plan_service.start_generation(db, str(user["_id"]), body)


@router.post("/replan", response_model=JobResponse, status_code=status.HTTP_202_ACCEPTED)
async def replan(
    db: AsyncIOMotorDatabase = Depends(get_db),
    user: dict = Depends(get_current_user),
):
    """Replan the weeks ahead, keeping everything through the week under way.

    The difference from `POST /plans` matters: that one rebuilds the plan from
    scratch, week one included, so a session already run and linked is rebuilt
    out from under the athlete. This one freezes the elapsed weeks on their
    original dates and regenerates only what is left, which keeps completed work
    — and the links made against it — exactly where they were.

    Takes no body on purpose: it reuses the active plan's objective. Adjusting
    how you get somewhere is not redefining where you are going; that is what
    the plan-setup screen is for."""
    try:
        return await plan_service.start_partial_replan(db, str(user["_id"]))
    except plan_service.NoActivePlanError:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail={"code": "NO_ACTIVE_PLAN", "message": "Aucun plan actif."},
        ) from None
    except plan_service.NothingLeftToReplanError:
        raise HTTPException(
            status_code=status.HTTP_409_CONFLICT,
            detail={
                "code": "NOTHING_TO_REPLAN",
                "message": (
                    "Ton plan se termine cette semaine : il n'y a plus de semaine à replanifier."
                ),
            },
        ) from None


@router.post("/replan-injury", response_model=JobResponse, status_code=status.HTTP_202_ACCEPTED)
async def replan_injury(
    body: InjuryReport,
    db: AsyncIOMotorDatabase = Depends(get_db),
    user: dict = Depends(get_current_user),
):
    """Declare an injury and replan the remaining weeks as a comeback (eased-off
    period then a gradual ramp). Reuses the current plan's objective; a plan must
    already exist. Queued like any generation — poll the returned job.

    Partial, like every other replan: the weeks through the one under way are
    frozen, so an injury declared on a Friday never rewrites the sessions already
    run and linked that week. What is left of the current week is handled by
    skipping sessions, which is what that override is for; the comeback phase
    starts with the first week actually being rewritten.
    """
    user_id = str(user["_id"])
    current = await plan_service.get_current_plan(db, user_id)
    if current is None or current.request is None:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail={"code": "NO_PLAN", "message": "Aucun plan à réadapter."},
        )
    try:
        base = await plan_service.build_replan_base(db, user_id)
    except plan_service.NothingLeftToReplanError:
        raise HTTPException(
            status_code=status.HTTP_409_CONFLICT,
            detail={
                "code": "NOTHING_TO_REPLAN",
                "message": (
                    "Ton plan se termine cette semaine : il n'y a plus de semaine à réadapter. "
                    "Passe les séances que tu ne peux pas faire."
                ),
            },
        ) from None
    return await plan_service.start_generation(db, user_id, current.request, injury=body, base=base)


@router.get("/today", response_model=TodaySession)
async def today_session(
    db: AsyncIOMotorDatabase = Depends(get_db),
    user: dict = Depends(get_current_user),
):
    """Today's planned session, deterministically adjusted for current form."""
    return await plan_service.get_today_session(db, str(user["_id"]))


@router.get("/progress", response_model=PlanProgress)
async def plan_progress_endpoint(
    db: AsyncIOMotorDatabase = Depends(get_db),
    user: dict = Depends(get_current_user),
):
    """Recent adherence + whether a replan is warranted (missed key sessions or
    persistent fatigue)."""
    return await plan_progress.compute_progress(db, str(user["_id"]))


@router.get("/reconcile", response_model=WeekReconciliation)
async def reconcile(
    db: AsyncIOMotorDatabase = Depends(get_db),
    user: dict = Depends(get_current_user),
):
    """Sessions of elapsed weeks the athlete has neither validated nor declared
    missed.

    Not an error state and never a 404: with no plan it answers `has_pending:
    false`. The app blocks on this, so it must not be able to block on the
    absence of a plan."""
    return await plan_reconcile_service.pending(db, str(user["_id"]))


@router.post("/reconcile/confirm-all")
async def reconcile_confirm_all(
    db: AsyncIOMotorDatabase = Depends(get_db),
    user: dict = Depends(get_current_user),
):
    """Link every pending session that came with a suggested activity."""
    return {"linked": await plan_reconcile_service.confirm_all_suggested(db, str(user["_id"]))}


@router.post("/reconcile/skip-all")
async def reconcile_skip_all(
    db: AsyncIOMotorDatabase = Depends(get_db),
    user: dict = Depends(get_current_user),
):
    """Declare every remaining pending session not done. Reversible, like any
    skip — the sessions stay in the plan, flagged."""
    return {"skipped": await plan_reconcile_service.resolve_all_missed(db, str(user["_id"]))}


@router.get("/session/link", response_model=SessionLinkInfo)
async def get_session_link(
    week_index: int,
    day: Weekday,
    slot: Literal["primary", "addon"] = "primary",
    db: AsyncIOMotorDatabase = Depends(get_db),
    user: dict = Depends(get_current_user),
):
    """The activity linked to a planned session (by week + weekday + slot), plus
    the session's real calendar date so the picker can surface nearby
    activities."""
    try:
        return await plan_completion_service.get_session_link(
            db, str(user["_id"]), week_index, day, slot
        )
    except plan_completion_service.NoActivePlanError as exc:
        raise _no_plan_error() from exc


@router.post("/session/link", response_model=SessionLinkInfo)
async def set_session_link(
    body: SessionLinkRequest,
    db: AsyncIOMotorDatabase = Depends(get_db),
    user: dict = Depends(get_current_user),
):
    """Link (or unlink, when activity_id is null) a recorded activity to a
    planned session."""
    try:
        return await plan_completion_service.set_session_link(
            db, str(user["_id"]), body.week_index, body.day, body.activity_id, body.slot
        )
    except plan_completion_service.NoActivePlanError as exc:
        raise _no_plan_error() from exc
    except plan_completion_service.ActivityNotFoundError as exc:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail={"code": "ACTIVITY_NOT_FOUND", "message": "Activité introuvable."},
        ) from exc
    except plan_completion_service.SportMismatchError as exc:
        raise HTTPException(
            status_code=status.HTTP_409_CONFLICT,
            detail={
                "code": "SPORT_MISMATCH",
                "message": (
                    "Cette séance est une séance de course : une activité "
                    f"{exc.activity_sport} ne peut pas la valider."
                ),
            },
        ) from exc


@router.post("/session/move", response_model=PlanResponse)
async def move_session(
    body: SessionMoveRequest,
    db: AsyncIOMotorDatabase = Depends(get_db),
    user: dict = Depends(get_current_user),
):
    """Move a session to another weekday for one week only (per-week override).
    Returns the updated current plan."""
    user_id = str(user["_id"])
    try:
        await plan_moves_service.move_session(
            db, user_id, body.week_index, body.from_day, body.to_day
        )
    except plan_moves_service.NoActivePlanError as exc:
        raise _no_plan_error() from exc
    except plan_moves_service.SessionNotFoundError as exc:
        raise _session_not_found_error() from exc
    plan = await plan_service.get_current_plan(db, user_id)
    if plan is None:
        raise _no_plan_error()
    return plan


@router.post("/session/duration", response_model=PlanResponse)
async def set_session_duration(
    body: SessionDurationRequest,
    db: AsyncIOMotorDatabase = Depends(get_db),
    user: dict = Depends(get_current_user),
):
    """Change a session's duration for one week only (per-week override).
    Returns the updated current plan."""
    user_id = str(user["_id"])
    try:
        await plan_moves_service.set_session_duration(
            db, user_id, body.week_index, body.day, body.slot, body.duration_min
        )
    except plan_moves_service.NoActivePlanError as exc:
        raise _no_plan_error() from exc
    except plan_moves_service.SessionNotFoundError as exc:
        raise _session_not_found_error() from exc
    plan = await plan_service.get_current_plan(db, user_id)
    if plan is None:
        raise _no_plan_error()
    return plan


@router.post("/session/skip", response_model=PlanResponse)
async def skip_session(
    body: SessionSkipRequest,
    db: AsyncIOMotorDatabase = Depends(get_db),
    user: dict = Depends(get_current_user),
):
    """Mark a session skipped, or un-skip it, for one week only (per-week
    override). Allowed on runs — unlike deletion — because it doesn't pretend
    the plan changed. It never triggers a regeneration: a skipped key session
    feeds `replan_suggested`, and acting on that stays the athlete's call."""
    user_id = str(user["_id"])
    try:
        await plan_moves_service.skip_session(
            db, user_id, body.week_index, body.day, body.slot, body.skipped
        )
    except plan_moves_service.NoActivePlanError as exc:
        raise _no_plan_error() from exc
    except plan_moves_service.SessionNotFoundError as exc:
        raise _session_not_found_error() from exc
    return await plan_service.get_current_plan(db, user_id)


@router.post("/session/delete", response_model=PlanResponse)
async def delete_session(
    body: SessionDeleteRequest,
    db: AsyncIOMotorDatabase = Depends(get_db),
    user: dict = Depends(get_current_user),
):
    """Drop a non-run session for one week only (per-week override). Returns the
    updated current plan."""
    user_id = str(user["_id"])
    try:
        await plan_moves_service.delete_session(db, user_id, body.week_index, body.day, body.slot)
    except plan_moves_service.NoActivePlanError as exc:
        raise _no_plan_error() from exc
    except plan_moves_service.SessionNotFoundError as exc:
        raise _session_not_found_error() from exc
    except plan_moves_service.CannotDeleteRunError as exc:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail={
                "code": "CANNOT_DELETE_RUN",
                "message": (
                    "Une séance de course ne se supprime pas à la semaine : "
                    "régénère le plan pour changer le nombre de courses."
                ),
            },
        ) from exc
    plan = await plan_service.get_current_plan(db, user_id)
    if plan is None:
        raise _no_plan_error()
    return plan


@router.post("/cancel", status_code=status.HTTP_204_NO_CONTENT)
async def cancel_plan(
    db: AsyncIOMotorDatabase = Depends(get_db),
    user: dict = Depends(get_current_user),
):
    """Stop the active plan. It leaves the current-plan reads (today, progress,
    per-week overrides) but stays readable in the version history."""
    try:
        await plan_service.cancel_current_plan(db, str(user["_id"]))
    except plan_service.NoActivePlanError as exc:
        raise _no_plan_error() from exc


@router.get("/current", response_model=PlanResponse)
async def current_plan(
    db: AsyncIOMotorDatabase = Depends(get_db),
    user: dict = Depends(get_current_user),
):
    plan = await plan_service.get_current_plan(db, str(user["_id"]))
    if plan is None:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail={"code": "NO_PLAN", "message": "Aucun plan pour l'instant."},
        )
    return plan


@router.get("/versions", response_model=list[PlanVersionSummary])
async def plan_versions(
    db: AsyncIOMotorDatabase = Depends(get_db),
    user: dict = Depends(get_current_user),
):
    """Read-only history of successful plan versions, newest first."""
    return await plan_service.list_plan_versions(db, str(user["_id"]))


@router.get("/versions/{version}", response_model=PlanResponse)
async def plan_version(
    version: int,
    db: AsyncIOMotorDatabase = Depends(get_db),
    user: dict = Depends(get_current_user),
):
    """A single past plan version, read-only."""
    plan = await plan_service.get_plan_version(db, str(user["_id"]), version)
    if plan is None:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail={"code": "NO_PLAN_VERSION", "message": "Version introuvable."},
        )
    return plan


@router.post("/versions/{version}/restore", response_model=PlanResponse)
async def restore_plan_version(
    version: int,
    db: AsyncIOMotorDatabase = Depends(get_db),
    user: dict = Depends(get_current_user),
):
    """Bring a past version back as the active plan.

    Non-destructive: the version is copied forward as a new one, so the plan it
    replaces stays in the history and a restore can itself be undone. This is
    the way back from a regeneration that rebuilt a week the athlete had already
    run — the case this endpoint exists for."""
    try:
        return await plan_service.restore_plan_version(db, str(user["_id"]), version)
    except plan_service.PlanVersionNotFoundError:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail={"code": "NO_PLAN_VERSION", "message": "Version introuvable."},
        ) from None


@router.get("/versions/{version}/overview", response_model=PlanOverview)
async def plan_version_overview(
    version: int,
    db: AsyncIOMotorDatabase = Depends(get_db),
    user: dict = Depends(get_current_user),
):
    """Calendar placement + week-by-week planned-vs-done key runs for one plan
    version — what the plan summary screen needs beyond the plan itself."""
    adherence = await plan_progress.compute_overview(db, str(user["_id"]), version)
    if adherence is None:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail={"code": "NO_PLAN_VERSION", "message": "Version introuvable."},
        )
    return adherence

File: mobile/src/lib/api/plans.ts
import type { Activity } from '@/lib/api/activities';
import { apiClient } from '@/lib/api/client';
import type { SportType } from '@/lib/api/types';

export type Weekday =
  | 'MONDAY'
  | 'TUESDAY'
  | 'WEDNESDAY'
  | 'THURSDAY'
  | 'FRIDAY'
  | 'SATURDAY'
  | 'SUNDAY';

export type FixedSport = { sport: SportType; day: Weekday; flexible: boolean };

export type StrengthPref = {
  enabled: boolean;
  sessions_per_week: number; // 1–2
  duration_min: number; // 10–45
};

export type PlanRequest = {
  goal_type: 'race' | 'distance' | 'fitness';
  distance_km: number | null;
  race_date: string | null; // ISO date
  target_time_min: number | null;
  /** Monday week 1 begins on (ISO date). Null lets the server choose: the
   * current week while it still has room to train, next Monday otherwise. */
  start_date: string | null;
  available_days: Weekday[];
  min_run_sessions_per_week: number;
  max_run_sessions_per_week: number;
  fixed_sports: FixedSport[];
  include_cross_training: boolean;
  strength: StrengthPref;
};

export type SessionType =
  | 'easy'
  | 'long_run'
  | 'tempo'
  | 'threshold'
  | 'intervals'
  | 'recovery'
  | 'cross_training'
  | 'strength'
  | 'test'
  | 'race'
  | 'rest';

export type PaceRange = { min_per_km_low: string; min_per_km_high: string };

export type PlanBlock = {
  label: string;
  duration_min: number;
  hr_zone: number | null;
  pace_range: PaceRange | null;
};

export type PlanSession = {
  day: Weekday;
  sport: SportType;
  type: SessionType;
  duration_min: number;
  structure: PlanBlock[];
  pace_range: PaceRange | null;
  hr_zone: number | null;
  priority: 'key' | 'optional';
  slot: 'primary' | 'addon';
  rationale: string;
  /** Declared skipped for this week (a per-week override, reversible). The
   * session stays in the plan — the decision is worth seeing. */
  skipped: boolean;
};

export type PlanWeek = {
  index: number;
  is_deload: boolean;
  target_load: number;
  sessions: PlanSession[];
};

export type PlanPhase = {
  name: 'base' | 'build' | 'peak' | 'taper';
  weeks: PlanWeek[];
};

export type Plan = {
  goal: { description: string; distance_km: number | null; race_date: string | null };
  phases: PlanPhase[];
};

export type PlanResponse = {
  id: string;
  status: 'generating' | 'ready' | 'failed';
  request: PlanRequest | null;
  plan: Plan | null;
  estimated_time_min: number | null; // estimated current time at the target
  /**
   * How far that estimate can be trusted, from the distance between the effort
   * it was extrapolated from and the target (Riegel drifts as the ratio grows).
   * It was computed server-side all along but never surfaced, so the app showed
   * a race time with the authority of a number the engine itself rated "low".
   */
  estimated_time_confidence: 'high' | 'medium' | 'low' | null;
  projected_time_min: number | null; // projected time at the plan's end
  feasibility_warning: string | null;
  error_message: string | null;
};

export type DailyAdjustment = {
  adjusted: boolean;
  original_type: SessionType;
  suggested_type: SessionType;
  reason: string;
};

/** Overnight recovery readings (Garmin) behind today's adjustment. */
export type RecoverySummary = {
  hrv: number | null;
  hrv_baseline: number | null;
  resting_hr: number | null;
  resting_hr_baseline: number | null;
  sleep_hours: number | null;
  body_battery: number | null;
  date: string | null;
};

export type TodaySession = {
  date: string;
  has_plan: boolean;
  has_session: boolean;
  week_index: number | null;
  sessions: PlanSession[];
  adjustment: DailyAdjustment | null;
  tsb: number;
  recovery: RecoverySummary | null;
  message: string | null;
};

export type PlanProgress = {
  has_plan: boolean;
  week_current: number | null;
  weeks_total: number | null;
  recent_key_planned: number;
  recent_key_completed: number;
  recent_key_missed: number;
  tsb: number;
  replan_suggested: boolean;
  replan_reason: string | null;
};

/** A user-declared injury — triggers a comeback replan. */
export type InjuryReport = {
  area: string;
  severity: 'gene' | 'douleur' | 'arret';
  days_off: number;
};

export function getPlanProgress() {
  return apiClient.get<PlanProgress>('/api/v1/plans/progress');
}

/**
 * Declare an injury and regenerate the remaining plan as a comeback (eased-off
 * period then a gradual ramp). Reuses the current plan's objective.
 */
/** Queues a comeback replan and returns its job — same polling as a fresh
 * generation (see `createPlan`). */
export function reportInjury(injury: InjuryReport) {
  return apiClient.post<PlanJob>('/api/v1/plans/replan-injury', injury);
}

export function getCurrentPlan() {
  return apiClient.get<PlanResponse>('/api/v1/plans/current');
}

/** One stored plan version, read-only. */
export type PlanVersionSummary = {
  version: number;
  created_at: string;
  status: 'ready' | 'cancelled';
  /** Whether this version is the plan currently driving the app. Comes from the
   * server: "the newest one" stopped being the rule once a plan could be
   * stopped, and re-deriving it here would just be a second place to get wrong. */
  is_active: boolean;
  goal_description: string | null;
  race_date: string | null;
  weeks_total: number | null;
  reason: string;
  injury_area: string | null;
  estimated_time_min: number | null;
  projected_time_min: number | null;
};

export function getPlanVersions() {
  return apiClient.get<PlanVersionSummary[]>('/api/v1/plans/versions');
}

export function getPlanVersion(version: number) {
  return apiClient.get<PlanResponse>(`/api/v1/plans/versions/${version}`);
}

/**
 * Bring a past version back as the active plan.
 *
 * Non-destructive: the version is copied forward as a new one, so the plan it
 * replaces stays in the history and a restore can itself be undone. This is the
 * way back from a regeneration that rebuilt a week already run.
 */
export function restorePlanVersion(version: number) {
  return apiClient.post<PlanResponse>(`/api/v1/plans/versions/${version}/restore`, {});
}

/** One week of a plan, committed-to versus actually run. Key runs only. */
export type WeekAdherence = {
  week_index: number;
  planned: number;
  completed: number;
  is_deload: boolean;
  /** A week is only scored once it is over — an unfinished week isn't "missed". */
  is_past: boolean;
  is_current: boolean;
};

/** What the plan summary screen needs beyond the plan document: where its weeks
 * fall on the calendar, and how much of it was actually run. Everything
 * derivable from the plan itself (volume, phases, sessions per week) is
 * computed client-side rather than sent twice. */
export type PlanOverview = {
  version: number;
  start_date: string; // ISO date
  end_date: string; // ISO date, inclusive
  week_current: number | null;
  weeks_total: number;
  planned: number; // totals over elapsed weeks only
  completed: number;
  weeks: WeekAdherence[];
};

export function getPlanOverview(version: number) {
  return apiClient.get<PlanOverview>(`/api/v1/plans/versions/${version}/overview`);
}

/**
 * Today's planned session, deterministically adjusted for current form (TSB).
 * No AI, no cost — the adjustment and its reason are computed server-side.
 */
export function getTodaySession() {
  return apiClient.get<TodaySession>('/api/v1/plans/today');
}

/**
 * Generate a new plan. Resolves only when the backend has generated, validated,
 * and stored it (synchronous, like the Garmin sync) — can take ~10-30s.
 */
/** A queued plan generation. */
export type PlanJob = {
  id: string;
  type: string;
  status: 'pending' | 'running' | 'done' | 'failed';
  /** When the job was queued (ISO, UTC). The server's clock is the only honest
   * source for "how long has this been running": the client's own start time is
   * lost the moment the PWA is reloaded mid-generation. */
  created_at: string;
  updated_at: string;
  error_message: string | null;
  result_summary: { plan_id?: string; status?: string } | null;
};

/**
 * Queues a generation and returns its job — it does NOT return a plan.
 *
 * A 17-week plan is ~12.5k output tokens, three to six minutes of model time;
 * no HTTP request should be held open for that. Poll `getPlanJob` until it
 * leaves `pending`/`running`, then read the plan from `getCurrentPlan`.
 * `usePlanGeneration` does all of that.
 */
export function createPlan(request: PlanRequest) {
  return apiClient.post<PlanJob>('/api/v1/plans', request);
}

/**
 * Replan what is left, keeping what has already been run.
 *
 * The weeks up to and including the one under way are frozen exactly as they
 * stand — sessions, moves, duration edits, links — and only the weeks after
 * them are regenerated, from current form. The plan's start date never moves,
 * so the week grid stays put.
 *
 * Takes no body on purpose: the objective comes from the active plan
 * server-side. Changing the objective is a different act, and it lives in
 * plan-setup (`createPlan`), which does rebuild everything.
 *
 * Same job protocol as `createPlan` — poll `getPlanJob`. 404 `NO_ACTIVE_PLAN`
 * when there is nothing to replan from, 409 `NOTHING_TO_REPLAN` when the plan
 * is on its last week.
 */
export function replanPlan() {
  return apiClient.post<PlanJob>('/api/v1/plans/replan', {});
}

export function getPlanJob(jobId: string) {
  return apiClient.get<PlanJob>(`/api/v1/jobs/${jobId}`);
}

/** The activity linked to a planned session, plus that session's calendar date
 * (so the picker can surface activities recorded near it). */
export type SessionLinkInfo = {
  session_date: string; // ISO date
  linked: Activity | null;
};

/** The slot is part of the identity: a day can hold a run and a strength addon,
 * and a link on one must not mark the other done. */
export function getSessionLink(weekIndex: number, day: Weekday, slot: PlanSession['slot'] = 'primary') {
  return apiClient.get<SessionLinkInfo>(
    `/api/v1/plans/session/link?week_index=${weekIndex}&day=${day}&slot=${slot}`,
  );
}

/** Link (activityId) or unlink (null) a recorded activity to a planned session. */
export function setSessionLink(
  weekIndex: number,
  day: Weekday,
  activityId: string | null,
  slot: PlanSession['slot'] = 'primary',
) {
  return apiClient.post<SessionLinkInfo>('/api/v1/plans/session/link', {
    week_index: weekIndex,
    day,
    slot,
    activity_id: activityId,
  });
}

/**
 * Stop the active plan. It leaves every current-plan read (today, progress,
 * per-week overrides) but stays readable in the version history, and a new plan
 * can be generated right after.
 */
export function cancelPlan() {
  return apiClient.post<void>('/api/v1/plans/cancel', {});
}

/**
 * Change a session's duration for one week only. Same override model as a move:
 * the stored plan is untouched and a replan clears it.
 */
export function setSessionDuration(
  weekIndex: number,
  day: Weekday,
  slot: PlanSession['slot'],
  durationMin: number,
) {
  return apiClient.post<PlanResponse>('/api/v1/plans/session/duration', {
    week_index: weekIndex,
    day,
    slot,
    duration_min: durationMin,
  });
}

/**
 * Declare a session skipped for one week only, or take it back. Allowed on runs
 * — unlike deleting one — because it says "not this week" rather than changing
 * the plan. A skipped key session counts as missed straight away and feeds
 * `replan_suggested`; it never regenerates anything on its own.
 */
export function skipSession(
  weekIndex: number,
  day: Weekday,
  slot: PlanSession['slot'],
  skipped: boolean,
) {
  return apiClient.post<PlanResponse>('/api/v1/plans/session/skip', {
    week_index: weekIndex,
    day,
    slot,
    skipped,
  });
}

/**
 * Drop a session for one week only. The backend refuses this for a run: losing
 * one would break the plan's guaranteed run count, which is a replan decision.
 */
export function deleteSession(weekIndex: number, day: Weekday, slot: PlanSession['slot']) {
  return apiClient.post<PlanResponse>('/api/v1/plans/session/delete', {
    week_index: weekIndex,
    day,
    slot,
  });
}

/**
 * Move a session to another weekday for one week only (a per-week override —
 * the stored plan is untouched). Returns the updated current plan.
 */
export function moveSession(weekIndex: number, fromDay: Weekday, toDay: Weekday) {
  return apiClient.post<PlanResponse>('/api/v1/plans/session/move', {
    week_index: weekIndex,
    from_day: fromDay,
    to_day: toDay,
  });
}
